File: ui/web_legacy_backup_20260504/src/lib/api/ai.js
import { BASE } from "./http.js";
export async function fetchStatus(provider = null, probe = true) {
	const params = new URLSearchParams();
	if (provider) params.set("provider", provider);
	if (!probe) params.set("probe", "0");
	const qs = params.toString();
	const res = await fetch(`${BASE}/api/status${qs ? `?${qs}` : ""}`);
	if (!res.ok) throw new Error("상태 확인 실패");
	return res.json();
}

export async function fetchAiSuggestions(stockCode) {
	const res = await fetch(`${BASE}/api/suggest?stockCode=${encodeURIComponent(stockCode)}`);
	if (!res.ok) {
		const err = await res.json().catch(() => ({}));
		throw new Error(err.detail || "추천 질문 조회 실패");
	}
	return res.json();
}

export async function validateProvider(provider, model = null, apiKey = null) {
	const body = { provider };
	if (model) body.model = model;
	if (apiKey) body.api_key = apiKey;
	const res = await fetch(`${BASE}/api/provider/validate`, {
		method: "POST",
		headers: { "Content-Type": "application/json" },
		body: JSON.stringify(body),
	});
	if (!res.ok) throw new Error("설정 실패");
	return res.json();
}

export const configure = validateProvider;

export async function fetchAiProfile() {
	const res = await fetch(`${BASE}/api/ai/profile`);
	if (!res.ok) throw new Error("AI profile 조회 실패");
	return res.json();
}

export async function updateAiProfile(patch) {
	const res = await fetch(`${BASE}/api/ai/profile`, {
		method: "PUT",
		headers: { "Content-Type": "application/json" },
		body: JSON.stringify(patch),
	});
	if (!res.ok) throw new Error("AI profile 저장 실패");
	return res.json();
}

export async function updateAiSecret(provider, apiKey = null, clear = false) {
	const res = await fetch(`${BASE}/api/ai/profile/secrets`, {
		method: "POST",
		headers: { "Content-Type": "application/json" },
		body: JSON.stringify({ provider, api_key: apiKey, clear }),
	});
	if (!res.ok) {
		const err = await res.json().catch(() => ({}));
		throw new Error(err.detail || "AI secret 저장 실패");
	}
	return res.json();
}

export async function validateDartKey(apiKey) {
	const res = await fetch(`${BASE}/api/openapi/dart-key/validate`, {
		method: "POST",
		headers: { "Content-Type": "application/json" },
		body: JSON.stringify({ api_key: apiKey }),
	});
	if (!res.ok) {
		const err = await res.json().catch(() => ({}));
		throw new Error(err.detail || "OpenDART 키 검증 실패");
	}
	return res.json();
}

export async function saveDartKey(apiKey) {
	const res = await fetch(`${BASE}/api/openapi/dart-key`, {
		method: "PUT",
		headers: { "Content-Type": "application/json" },
		body: JSON.stringify({ api_key: apiKey }),
	});
	if (!res.ok) {
		const err = await res.json().catch(() => ({}));
		throw new Error(err.detail || "OpenDART 키 저장 실패");
	}
	return res.json();
}

export async function deleteDartKey() {
	const res = await fetch(`${BASE}/api/openapi/dart-key`, { method: "DELETE" });
	if (!res.ok) {
		const err = await res.json().catch(() => ({}));
		throw new Error(err.detail || "OpenDART 키 삭제 실패");
	}
	return res.json();
}

export function subscribeAiProfileEvents({ onProfileChanged, onError } = {}) {
	const es = new EventSource(`${BASE}/api/ai/profile/events`);
	es.addEventListener("profile_changed", (event) => {
		try {
			const payload = JSON.parse(event.data);
			onProfileChanged?.(payload);
		} catch (err) {
			console.warn("profile_changed parse:", err);
		}
	});
	es.onerror = (err) => {
		onError?.(err);
	};
	return {
		close() { es.close(); },
	};
}

export async function fetchModels(provider) {
	const res = await fetch(`${BASE}/api/models/${encodeURIComponent(provider)}`);
	if (!res.ok) return { models: [] };
	return res.json();
}

export function pullOllamaModel(modelName, { onProgress, onDone, onError }) {
	const controller = new AbortController();
	fetch(`${BASE}/api/ollama/pull`, {
		method: "POST",
		headers: { "Content-Type": "application/json" },
		body: JSON.stringify({ model: modelName }),
		signal: controller.signal,
	})
	.then(async (res) => {
		if (!res.ok) { onError?.("다운로드 실패"); return; }
		const reader = res.body.getReader();
		const decoder = new TextDecoder();
		let buffer = "";
		while (true) {
			const { done, value } = await reader.read();
			if (done) break;
			buffer += decoder.decode(value, { stream: true });
			const lines = buffer.split("\n");
			buffer = lines.pop() || "";
			for (const line of lines) {
				if (line.startsWith("data:")) {
					try {
						const data = JSON.parse(line.slice(5).trim());
						if (data.total && data.completed !== undefined) {
							onProgress?.({ total: data.total, completed: data.completed, status: data.status });
						} else if (data.status) {
							onProgress?.({ status: data.status });
						}
					} catch (e) { console.warn("SSE parse:", e); }
				}
			}
		}
		onDone?.();
	})
	.catch((err) => {
		if (err.name !== "AbortError") onError?.(err.message);
	});
	return { abort: () => controller.abort() };
}

export async function fetchDevChannelStatus() {
	const res = await fetch(`${BASE}/api/channel`);
	if (!res.ok) throw new Error("Channel 상태 확인 실패");
	return res.json();
}

export async function startDevChannel() {
	const res = await fetch(`${BASE}/api/channel/start`, { method: "POST" });
	if (!res.ok) {
		const err = await res.json().catch(() => ({}));
		throw new Error(err.detail || "Channel 시작 실패");
	}
	return res.json();
}

export async function stopDevChannel() {
	const res = await fetch(`${BASE}/api/channel/stop`, { method: "POST" });
	if (!res.ok) {
		const err = await res.json().catch(() => ({}));
		throw new Error(err.detail || "Channel 종료 실패");
	}
	return res.json();
}

export async function codexLogout() {
	const res = await fetch(`${BASE}/api/codex/logout`, { method: "POST" });
	if (!res.ok) throw new Error("Codex 로그아웃 실패");
	return res.json();
}

export async function oauthAuthorize() {
	const res = await fetch(`${BASE}/api/oauth/authorize`);
	if (!res.ok) throw new Error("OAuth 로그인 시작 실패");
	return res.json();
}

export async function oauthStatus() {
	const res = await fetch(`${BASE}/api/oauth/status`);
	if (!res.ok) throw new Error("OAuth 상태 확인 실패");
	return res.json();
}

export async function oauthLogout() {
	const res = await fetch(`${BASE}/api/oauth/logout`, { method: "POST" });
	if (!res.ok) throw new Error("OAuth 로그아웃 실패");
	return res.json();
}

export async function startChannelConnection(platform, payload = {}) {
	const res = await fetch(`${BASE}/api/channels/${encodeURIComponent(platform)}/start`, {
		method: "POST",
		headers: { "Content-Type": "application/json" },
		body: JSON.stringify(payload),
	});
	if (!res.ok) {
		const err = await res.json().catch(() => ({}));
		throw new Error(err.detail || `${platform} 채널 시작 실패`);
	}
	return res.json();
}

export async function stopChannelConnection(platform) {
	const res = await fetch(`${BASE}/api/channels/${encodeURIComponent(platform)}/stop`, {
		method: "POST",
	});
	if (!res.ok) {
		const err = await res.json().catch(() => ({}));
		throw new Error(err.detail || `${platform} 채널 종료 실패`);
	}
	return res.json();
}

File: src/dartlab/server/api/ai.py
from __future__ import annotations

import asyncio
import json
import logging
import os
from typing import Any

from fastapi import APIRouter, HTTPException, Query, Request
from sse_starlette.sse import EventSourceResponse

import dartlab
from dartlab.ai.settings import (
    build_provider_catalog,
    get_profile_manager,
    get_provider_spec,
    public_provider_ids,
)
from dartlab.ai.settings.modelResolver import fallback_models, is_openai_chat_model, sort_openai_models

from ..chat import OLLAMA_MODEL_GUIDE
from ..models import (
    AiProfileUpdateRequest,
    AiSecretUpdateRequest,
    ChannelConnectRequest,
    ConfigureRequest,
    DartKeyUpdateRequest,
)
from ..services.aiProfile import (
    build_codex_detail,
    build_oauth_codex_detail,
    build_ollama_detail,
    probe_provider_availability,
    validate_provider_connection,
)
from .common import (
    HANDLED_API_ERRORS as _HANDLED_API_ERRORS,
)
from .common import (
    guideDetail as _guideDetail,
)
from .common import (
    normalize_provider_name as _normalize_provider_name,
)

logger = logging.getLogger(__name__)

router = APIRouter()

UI_PROVIDERS = public_provider_ids()
STATIC_MODELS: dict[str, list[str]] = {}

_oauth_state: dict[str, Any] = {}


def _build_open_dart_status() -> dict[str, Any]:
    from dartlab.providers.dart.openapi.dartKey import getDartKeyStatus

    return getDartKeyStatus().toDict()


def _resolve_credential_source(meta: dict[str, Any], profile_provider: dict[str, Any]) -> str:
    auth_kind = meta.get("authKind", "none")
    if auth_kind == "api_key":
        if profile_provider.get("secretConfigured"):
            return "secret_store"
        env_key = meta.get("envKey")
        if env_key and os.environ.get(env_key):
            return "env"
        return "none"
    if auth_kind == "oauth":
        return "oauth" if profile_provider.get("secretConfigured") else "none"
    if auth_kind == "cli":
        return "cli"
    return "none"


@router.get("/api/status")
def api_status(
    provider: str | None = Query(None, description="상태를 적극 확인할 provider"),
    probe: bool = Query(True, description="True면 provider availability를 실제 점검"),
):
    """LLM provider 상태 확인 (설치/인증/모델 포함)."""
    profile_snapshot = get_profile_manager().serialize()
    catalog = {item["id"]: item for item in build_provider_catalog()}
    results = {}
    target_provider = _normalize_provider_name(provider) or provider
    if probe and target_provider is None:
        target_provider = _normalize_provider_name(profile_snapshot.get("defaultProvider")) or profile_snapshot.get(
            "defaultProvider"
        )
    role_bindings = profile_snapshot.get("roles", {})

    for prov in UI_PROVIDERS:
        meta = catalog.get(prov, {})
        profile_provider = profile_snapshot.get("providers", {}).get(prov, {})
        info: dict[str, Any] = {
            "available": None,
            "model": profile_provider.get("model"),
            "checked": False,
            "label": meta.get("label", prov),
            "desc": meta.get("description", ""),
            "auth": meta.get("authKind", "none"),
            "secretConfigured": bool(profile_provider.get("secretConfigured")),
            "credentialSource": _resolve_credential_source(meta, profile_provider),
            "selected": profile_snapshot.get("defaultProvider") == prov,
            "selectedRoles": [
                role_name
                for role_name, binding in role_bindings.items()
                if isinstance(binding, dict) and binding.get("provider") == prov
            ],
        }
        if meta.get("envKey"):
            info["envKey"] = meta["envKey"]
        if meta.get("signupUrl"):
            info["signupUrl"] = meta["signupUrl"]
        if meta.get("freeTierHint"):
            info["freeTierHint"] = meta["freeTierHint"]
        should_probe = probe and (target_provider is None or prov == target_provider)
        if should_probe:
            available, model, checked = probe_provider_availability(prov)
            info["available"] = available
            info["model"] = model
            info["checked"] = checked
        else:
            # probe 안 했으면 secretConfigured 로 fallback — UI 가 null 을 "검증 실패" 로 잘못
            # 해석해서 "설정 필요" 표시하던 문제 차단. probe 결과는 별도로 (백그라운드 또는
            # Settings 패널 진입 시) 갱신.
            info["available"] = info["secretConfigured"]
        results[prov] = info

    ollama_detail = build_ollama_detail(probe=probe and (target_provider is None or target_provider == "ollama"))
    oauth_codex_detail = build_oauth_codex_detail(
        probe=probe and (target_provider is None or target_provider == "oauth-codex")
    )
    codex_detail = build_codex_detail(probe=probe and (target_provider is None or target_provider == "codex"))

    version = dartlab.__version__ if hasattr(dartlab, "__version__") else "unknown"

    # Room 정보 (터널 모드에서 협업 세션 활성 시)
    room_info = None
    try:
        from ..room import room_manager

        active_room = room_manager.get_room()
        if active_room is not None:
            room_info = {
                "roomId": active_room.room_id,
                "members": len(active_room.members),
            }
    except ImportError:
        pass

    resp: dict[str, Any] = {
        "providers": results,
        "ollama": ollama_detail,
        "codex": codex_detail,
        "oauthCodex": oauth_codex_detail,
        "openDart": _build_open_dart_status(),
        "profile": profile_snapshot,
        "version": version,
    }
    if room_info is not None:
        resp["room"] = room_info
    try:
        from ..services.channelRuntime import channel_runtime

        resp["channels"] = channel_runtime.status()
    except ImportError:
        resp["channels"] = {}
    try:
        from ..services.devChannelRuntime import dev_channel_runtime

        resp["channel"] = dev_channel_runtime.status()
    except ImportError:
        resp["channel"] = {"kind": "devtunnel", "running": False, "url": None, "qrDataUrl": None, "error": None}
    return resp


@router.get("/api/suggest")
def api_suggest(stockCode: str = Query(..., description="추천 질문을 생성할 종목코드")):
    """회사 데이터 상태에 맞는 추천 질문 목록을 반환한다."""
    try:
        from ..services.companyApi import get_company

        company = get_company(stockCode)
        return {
            "stockCode": getattr(company, "stockCode", stockCode),
            "company": getattr(company, "corpName", stockCode),
            "suggestions": [],
            "dataReady": {},
        }
    except ValueError as e:
        raise HTTPException(status_code=400, detail=_guideDetail(e)) from e
    except _HANDLED_API_ERRORS as e:
        raise HTTPException(status_code=500, detail=_guideDetail(e)) from e


@router.post("/api/provider/validate")
def api_validate_provider(req: ConfigureRequest):
    """LLM provider 검증. 전역 상태는 변경하지 않는다."""
    return validate_provider_connection(req)


@router.post("/api/configure")
def api_configure(req: ConfigureRequest):
    """구버전 alias. 현재는 provider 검증만 수행한다."""
    return validate_provider_connection(req)


@router.get("/api/ai/profile")
def api_ai_profile():
    """공통 AI profile + provider catalog 반환."""
    return get_profile_manager().serialize()


@router.put("/api/ai/profile")
def api_ai_profile_update(req: AiProfileUpdateRequest):
    """공통 AI profile 갱신."""
    from dartlab.ai import configure as configure_ai

    manager = get_profile_manager()
    provider = _normalize_provider_name(req.provider) or req.provider
    if provider and get_provider_spec(provider) is None:
        raise HTTPException(status_code=400, detail=f"지원하지 않는 provider: {provider}")
    profile = manager.update(
        provider=provider,
        role=req.role,
        model=req.model,
        base_url=req.base_url,
        temperature=req.temperature,
        max_tokens=req.max_tokens,
        system_prompt=req.system_prompt,
        updated_by="ui",
    )
    if provider:
        configure_ai(
            provider=provider,
            role=req.role,
            model=req.model,
            base_url=req.base_url,
            temperature=req.temperature,
            max_tokens=req.max_tokens,
            system_prompt=req.system_prompt,
        )
    return manager.serialize() | {"revision": profile.revision}


@router.post("/api/ai/profile/secrets")
def api_ai_profile_secret(req: AiSecretUpdateRequest):
    """provider secret 저장/삭제."""
    provider = _normalize_provider_name(req.provider) or req.provider
    spec = get_provider_spec(provider)
    if spec is None:
        raise HTTPException(status_code=400, detail=f"지원하지 않는 provider: {provider}")
    if spec.auth_kind != "api_key":
        raise HTTPException(status_code=400, detail=f"{provider} provider는 API key secret을 사용하지 않습니다")

    manager = get_profile_manager()
    if req.clear or not req.api_key:
        profile = manager.clear_api_key(provider, updated_by="ui")
    else:
        profile = manager.save_api_key(provider, req.api_key, updated_by="ui")
    return manager.serialize() | {"revision": profile.revision}


@router.post("/api/openapi/dart-key/validate")
def api_validate_dart_key(req: DartKeyUpdateRequest):
    """OpenDART API 키 유효성만 검증한다."""
    from dartlab.providers.dart.openapi.dartKey import validateDartApiKey

    api_key = (req.api_key or "").strip()
    if not api_key:
        raise HTTPException(status_code=400, detail="DART API 키를 입력하세요.")
    try:
        result = validateDartApiKey(api_key)
        return result | {"openDart": _build_open_dart_status()}
    except ValueError as e:
        raise HTTPException(status_code=400, detail=_guideDetail(e)) from e
    except _HANDLED_API_ERRORS as e:
        raise HTTPException(status_code=500, detail=_guideDetail(e)) from e


@router.put("/api/openapi/dart-key")
def api_save_dart_key(req: DartKeyUpdateRequest):
    """프로젝트 .env에 OpenDART API 키를 저장한다."""
    from dartlab.providers.dart.openapi.dartKey import saveDartKeyToDotenv

    api_key = (req.api_key or "").strip()
    if not api_key:
        raise HTTPException(status_code=400, detail="DART API 키를 입력하세요.")
    try:
        env_path = saveDartKeyToDotenv(api_key)
        return {"ok": True, "envPath": str(env_path), "openDart": _build_open_dart_status()}
    except OSError as e:
        raise HTTPException(status_code=500, detail=_guideDetail(e)) from e


@router.delete("/api/openapi/dart-key")
def api_delete_dart_key():
    """프로젝트 .env의 OpenDART API 키를 제거한다."""
    from dartlab.providers.dart.openapi.dartKey import clearDartKeyFromDotenv

    try:
        env_path = clearDartKeyFromDotenv()
        return {"ok": True, "envPath": str(env_path), "openDart": _build_open_dart_status()}
    except OSError as e:
        raise HTTPException(status_code=500, detail=_guideDetail(e)) from e


@router.post("/api/channels/{platform}/start")
def api_channel_start(platform: str, req: ChannelConnectRequest):
    """외부 채널 어댑터 시작."""
    try:
        from ..services.channelRuntime import channel_runtime

        payload = req.model_dump(exclude_none=True)
        return channel_runtime.start(platform, **payload)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=_guideDetail(e)) from e
    except _HANDLED_API_ERRORS as e:
        raise HTTPException(status_code=500, detail=_guideDetail(e)) from e


@router.post("/api/channels/{platform}/stop")
def api_channel_stop(platform: str):
    """외부 채널 어댑터 정지."""
    try:
        from ..services.channelRuntime import channel_runtime

        return channel_runtime.stop(platform)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=_guideDetail(e)) from e
    except _HANDLED_API_ERRORS as e:
        raise HTTPException(status_code=500, detail=_guideDetail(e)) from e


def _request_port(request: Request) -> int:
    if request.url.port:
        return int(request.url.port)
    return 8400


@router.get("/api/channel")
def api_dev_channel_status():
    """DevTunnels 모바일 접속 채널 상태를 반환한다."""
    try:
        from ..services.devChannelRuntime import dev_channel_runtime

        return dev_channel_runtime.status()
    except _HANDLED_API_ERRORS as e:
        raise HTTPException(status_code=500, detail=_guideDetail(e)) from e


@router.post("/api/channel/start")
def api_dev_channel_start(request: Request):
    """현재 Web UI를 모바일에서 열 수 있는 DevTunnels 채널을 시작한다."""
    try:
        from ..services.devChannelRuntime import dev_channel_runtime

        return dev_channel_runtime.start(port=_request_port(request), auto_yes=True)
    except _HANDLED_API_ERRORS as e:
        raise HTTPException(status_code=500, detail=_guideDetail(e)) from e


@router.post("/api/channel/stop")
def api_dev_channel_stop():
    """DevTunnels 채널을 종료한다."""
    try:
        from ..services.devChannelRuntime import dev_channel_runtime

        return dev_channel_runtime.stop()
    except _HANDLED_API_ERRORS as e:
        raise HTTPException(status_code=500, detail=_guideDetail(e)) from e


@router.get("/api/ai/profile/events")
async def api_ai_profile_events(request: Request):
    """profile 변경 SSE 스트림."""
    manager = get_profile_manager()

    async def _generate():
        last_fingerprint = ""
        while True:
            if await request.is_disconnected():
                break
            payload = manager.serialize()
            fingerprint = manager.fingerprint()
            if fingerprint != last_fingerprint:
                last_fingerprint = fingerprint
                yield {
                    "event": "profile_changed",
                    "data": json.dumps(payload, ensure_ascii=False),
                }
            await asyncio.sleep(1.0)

    return EventSourceResponse(_generate())


@router.get("/api/models/{provider}")
def api_models(provider: str):
    """Provider별 사용 가능한 모델 목록 — SDK/API 자동 조회, 실패시 fallback."""
    from dartlab.ai.providers import create_provider
    from dartlab.ai.settings.types import LLMConfig

    provider = _normalize_provider_name(provider) or provider

    if provider == "codex":
        try:
            from dartlab.ai.providers.support.codexCli import get_codex_model_catalog

            return {"models": get_codex_model_catalog()}
        except (ImportError, OSError, RuntimeError, ValueError):
            return {"models": fallback_models("codex", allow_fetch=False)}

    if provider == "oauth-codex":
        try:
            from dartlab.ai.providers.oauthCodex import availableModels

            # cache 우선 — 비어 있으면 정적 fallback 즉시 반환 + background thread 에서 warm.
            # 이전: cold 1 회 ~43s (DNS/TLS cold + remote /codex/models fetch) 동안 UI 가
            # "설정 필요" 표시. allow_fetch=False 로 fallback 도 cold HTTP 안 트리거.
            cached = availableModels(allow_fetch=False)
            if cached:
                return {"models": cached}
            import threading

            threading.Thread(target=availableModels, daemon=True).start()
            return {"models": fallback_models("oauth-codex", allow_fetch=False)}
        except (ImportError, OSError, RuntimeError, ValueError):
            return {"models": fallback_models("oauth-codex", allow_fetch=False)}

    if provider in STATIC_MODELS:
        return {"models": STATIC_MODELS[provider]}

    if provider == "ollama":
        try:
            config = LLMConfig(provider="ollama")
            prov = create_provider(config)
            installed = prov.get_installed_models()
            return {"models": installed, "recommendations": OLLAMA_MODEL_GUIDE}
        except _HANDLED_API_ERRORS:
            return {"models": [], "recommendations": OLLAMA_MODEL_GUIDE}

    if provider == "openai":
        models = _fetch_openai_models()
        if models:
            return {"models": models}
        return {"models": fallback_models("openai", allow_fetch=False)}

    return {"models": []}


def _get_api_key(provider: str) -> str | None:
    """글로벌 config 또는 환경변수에서 API 키를 가져온다."""
    from dartlab.ai import get_config

    config = get_config(provider)
    if config.api_key:
        return config.api_key
    env_map = {"openai": "OPENAI_API_KEY"}
    return os.environ.get(env_map.get(provider, ""))


def _fetch_openai_models() -> list[str]:
    """OpenAI SDK로 모델 목록을 가져온다."""
    api_key = _get_api_key("openai")
    if not api_key:
        return []
    try:
        from openai import OpenAI

        client = OpenAI(api_key=api_key)
        raw = client.models.list()
        models = []
        for model in raw:
            mid = model.id
            if is_openai_chat_model(mid):
                models.append(mid)
        return sort_openai_models(models)
    except (ImportError, OSError, RuntimeError, ValueError):
        return []


@router.post("/api/codex/logout")
def api_codex_logout():
    """Codex CLI에 저장된 계정 인증을 제거한다."""
    try:
        from dartlab.ai.providers.support.codexCli import logout_codex_cli

        logout_codex_cli()
    except ImportError:
        return {"ok": True}
    except FileNotFoundError as exc:
        raise HTTPException(status_code=404, detail=_guideDetail(exc)) from exc
    except RuntimeError as exc:
        raise HTTPException(status_code=400, detail=_guideDetail(exc)) from exc
    return {"ok": True}


@router.get("/api/oauth/authorize")
def api_oauth_authorize():
    """ChatGPT OAuth 인증 시작 — 브라우저 로그인 URL 반환 + 로컬 콜백 서버 시작."""
    from dartlab.ai.providers.support.oauthToken import OAUTH_REDIRECT_PORT, build_auth_url

    auth_url, verifier, state = build_auth_url()

    _oauth_state["verifier"] = verifier
    _oauth_state["state"] = state
    _oauth_state["done"] = False
    _oauth_state["error"] = None

    _start_oauth_callback_server(OAUTH_REDIRECT_PORT)

    return {"authUrl": auth_url, "state": state}


@router.get("/api/oauth/status")
def api_oauth_status():
    """OAuth 인증 완료 여부 폴링."""
    if _oauth_state.get("error"):
        return {"done": True, "error": _oauth_state["error"]}
    if _oauth_state.get("done"):
        return {"done": True, "error": None}
    return {"done": False}


@router.post("/api/oauth/logout")
def api_oauth_logout():
    """OAuth 토큰 제거."""
    try:
        from dartlab.ai.providers.support.oauthToken import revoke_token

        revoke_token()
    except (ImportError, OSError, RuntimeError, ValueError):
        pass
    get_profile_manager().update(provider="oauth-codex", updated_by="ui")
    return {"ok": True}


def _start_oauth_callback_server(port: int):
    """OAuth 콜백을 받을 임시 HTTP 서버를 백그라운드 스레드로 시작."""
    import threading
    from http.server import BaseHTTPRequestHandler, HTTPServer
    from urllib.parse import parse_qs, urlparse

    class CallbackHandler(BaseHTTPRequestHandler):
        def do_GET(self):
            parsed = urlparse(self.path)
            if parsed.path != "/auth/callback":
                self.send_response(404)
                self.end_headers()
                return

            params = parse_qs(parsed.query)
            code = (params.get("code") or [None])[0]
            state = (params.get("state") or [None])[0]
            error = (params.get("error") or [None])[0]

            if error:
                _oauth_state["error"] = error
                _oauth_state["done"] = True
                self._respond_html("인증 실패", f"오류: {error}")
                return

            if state != _oauth_state.get("state"):
                _oauth_state["error"] = "state_mismatch"
                _oauth_state["done"] = True
                self._respond_html("인증 실패", "보안 검증 실패 (state mismatch)")
                return

            if not code:
                _oauth_state["error"] = "no_code"
                _oauth_state["done"] = True
                self._respond_html("인증 실패", "인증 코드를 받지 못했습니다")
                return

            try:
                from dartlab.ai.providers.support.oauthToken import exchange_code

                exchange_code(code, _oauth_state["verifier"])
                get_profile_manager().update(provider="oauth-codex", updated_by="ui")
                _oauth_state["done"] = True
                self._respond_html("인증 성공", "DartLab 인증이 완료되었습니다. 이 창을 닫아주세요.")
            except _HANDLED_API_ERRORS as exc:
                _oauth_state["error"] = str(exc)
                _oauth_state["done"] = True
                self._respond_html("인증 실패", f"토큰 교환 실패: {exc}")

        def _respond_html(self, title: str, message: str):
            import html as _html

            safe_title = _html.escape(title)
            safe_message = _html.escape(message)
            markup = (
                "<!DOCTYPE html><html><head><meta charset='utf-8'>"
                f"<title>{safe_title}</title>"
                "<style>body{font-family:system-ui;display:flex;align-items:center;"
                "justify-content:center;min-height:100vh;margin:0;background:#050811;color:#e5e5e5}"
                "div{text-align:center;padding:2rem}"
                "h1{font-size:1.5rem;margin-bottom:1rem}"
                "</style></head><body>"
                f"<div><h1>{safe_title}</h1><p>{safe_message}</p></div>"
                "<script>setTimeout(()=>window.close(),3000)</script>"
                "</body></html>"
            )
            self.send_response(200)
            self.send_header("Content-Type", "text/html; charset=utf-8")
            self.end_headers()
            self.wfile.write(markup.encode("utf-8"))

        def log_message(self, fmt, *args):
            pass

    def _run_server():
        server = HTTPServer(("127.0.0.1", port), CallbackHandler)
        server.timeout = 120
        server.handle_request()
        server.server_close()

    thread = threading.Thread(target=_run_server, daemon=True)
    thread.start()


@router.post("/api/ollama/pull")
async def api_ollama_pull(req: dict):
    """Ollama 모델 다운로드 (SSE 스트리밍 진행률)."""
    model_name = req.get("model")
    if not model_name:
        raise HTTPException(400, "model name required")

    async def _stream_pull():
        import httpx

        try:
            with httpx.Client(timeout=600) as client:
                with client.stream(
                    "POST",
                    "http://localhost:11434/api/pull",
                    json={"model": model_name, "stream": True},
                ) as resp:
                    for line in resp.iter_lines():
                        if line:
                            yield {
                                "event": "progress",
                                "data": line,
                            }
            yield {"event": "done", "data": "{}"}
        except _HANDLED_API_ERRORS as exc:
            yield {"event": "error", "data": json.dumps({"error": _guideDetail(exc)}, ensure_ascii=False)}

    return EventSourceResponse(_stream_pull(), media_type="text/event-stream")
